File: 290. Word Pattern.cpp
//290. Word Pattern.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <sstream>
using namespace std;

// Created by bryantbyr on 20171012
// Time: O(n)
// Space: O(n)
// Hash Table
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        unordered_map<char, int> m1;
        unordered_map<string, int> m2;
        istringstream is(str);
        string s;
        for (unsigned i = 0; i < pattern.size(); ++i) {
            if (!(is >> s))
                return false;
            if (m1.find(pattern[i]) == m1.end() && m2.find(s) == m2.end())
                m1[pattern[i]] = m2[s] = i;
            else if (m1.find(pattern[i]) != m1.end() && m2.find(s) != m2.end() && m1[pattern[i]] == m2[s])
                continue;
            else
                return false;
        }
        // if (is >> s)
        //     return false;
        // return true;
        return !(is >> s);
    }
};

// Learn from discuss on 20171012
// Time: O(n)
// Space: O(n)
// Hash Table (initialize)
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        unordered_map<char, int> m1;
        unordered_map<string, int> m2;
        istringstream is(str);
        string s;
        for (unsigned i = 0; i < pattern.size(); ++i) {
            if (!(is >> s) || m1[pattern[i]] != m2[s])
                return false;
            // if (m1.find(pattern[i]) == m1.end() && m2.find(s) == m2.end())
            //     m1[pattern[i]] = m2[s] = i;
            m1[pattern[i]] = m2[s] = i + 1;
        }
        return !(is >> s);
    }
};

//Learn from discuss on 20171012
//Time:O(n)
//Space:O(n)
//Hash Table (initialize)
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        unordered_map<char, int> m1;
        unordered_map<string, int> m2;
        istringstream in(str);
        int i = 0, n = pattern.size();
        for (string word; in >> word; ++i) {
            if (i == n || m1[pattern[i]] != m2[word])
                return false;
            m1[pattern[i]] = m2[word] = i + 1;
        }
        return i == n;
    }
};

int main()
{
    Solution S;
    string s = "abba";
    string t = "dog cat cat dog dog";
    cout << S.wordPattern(s, t) << endl;
}

File: _290_Word_Pattern.java
//_290_Word_Pattern.java

import java.util.HashMap;
import java.util.Map;

//Created by bryantbyr on 20171012
//Time:O(n)
//Space:O(n)
//Hash Table
class Solution {
    public boolean wordPattern(String pattern, String str) {
        String[] s = str.split(" ");
        if (s.length != pattern.length())
            return false;
        Map map = new HashMap<>();
        for (Integer i = 0; i < s.length; ++i ) {
            if (map.put(pattern.charAt(i), i) != map.put(s[i], i))
                return false;
        }
        return true;
    }
}

//Created by bryantbyr on 20171012
//Time:O(n)
//Space:O(n)
//Hash Table
class Solution {
    public boolean wordPattern(String pattern, String str) {
        String[] s = str.split(" ");
        if (s.length != pattern.length())
            return false;
        Map<Character, Integer> m1 = new HashMap<>();
        Map<String, Integer> m2 = new HashMap<>();
        for (Integer i = 0; i < s.length; ++i ) {
            if (m1.get(pattern.charAt(i)) != m2.get(s[i]))
                return false;
            m1.put(pattern.charAt(i), i);
            m2.put(s[i], i);
        }
        return true;
    }
}



public class _290_Word_Pattern {
    public static void main(String[] args) {
        Solution S = new Solution();
        System.out.println(S.wordPattern("abba", "dog cat cat dog"));
    }
}
